File: java/src/test/java/org/oojs/SubtypeCheckTest.java
package org.oojs;

import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import static org.junit.jupiter.api.Assertions.*;

class SubtypeCheckTest extends TestHelpers {

    @Test void direct_subtype() {
        Schema schema = (Schema) animalRegistry()[1];
        assertTrue(schema.types.get("Dog").isSubtypeOf(schema.types.get("Animal")));
    }

    @Test void same_type() {
        Schema schema = (Schema) animalRegistry()[1];
        assertTrue(schema.types.get("Dog").isSubtypeOf(schema.types.get("Dog")));
    }

    @Test void not_subtype() {
        Schema schema = (Schema) animalRegistry()[1];
        assertFalse(schema.types.get("Cat").isSubtypeOf(schema.types.get("Dog")));
    }

    @Test void transitive() {
        Registry r = new Registry();
        r.loadMap(map("$oojs", "1.0", "$id", "https://example.org/schemas/deep",
                "types", map(
                        "A", map("abstract", true),
                        "B", map("extends", "A"),
                        "C", map("extends", "B"),
                        "D", map("extends", "C"))));
        Schema schema = r.getSchema("https://example.org/schemas/deep");
        assertTrue(schema.types.get("D").isSubtypeOf(schema.types.get("A")));
    }
}

File: php/tests/SubtypeCheckTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\Registry;
use PHPUnit\Framework\TestCase;

/**
 * §9.4 — Subtype check (isSubtypeOf).
 */
class SubtypeCheckTest extends TestCase
{
    use TestHelpers;

    public function test_direct_subtype(): void
    {
        [$r, $schema] = self::animalRegistry();
        $this->assertTrue($schema->types['Dog']->isSubtypeOf($schema->types['Animal']));
    }

    public function test_same_type(): void
    {
        [$r, $schema] = self::animalRegistry();
        $this->assertTrue($schema->types['Dog']->isSubtypeOf($schema->types['Dog']));
    }

    public function test_not_subtype(): void
    {
        [$r, $schema] = self::animalRegistry();
        $this->assertFalse($schema->types['Cat']->isSubtypeOf($schema->types['Dog']));
    }

    public function test_transitive(): void
    {
        $r = new Registry();
        $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/deep',
            'types' => [
                'A' => ['abstract' => true],
                'B' => ['extends' => 'A'],
                'C' => ['extends' => 'B'],
                'D' => ['extends' => 'C'],
            ],
        ]);
        $schema = $r->getSchema('https://example.org/schemas/deep');
        $this->assertTrue($schema->types['D']->isSubtypeOf($schema->types['A']));
    }
}
